### DISCUSSION TITLE
[Serializer] DX friendly way to ensure that a scalar datetime field has valid format when normalizing

### DISCUSSION THREAD

**Discussion by ghost** [symfony/symfony#53601] [Q&A]
Let's say I have an object or an array key with a property/key that represents a datetime but is of type string (e.g. data received directly from database):
```php
class A 
{
  public string $createdAt = '2024-23-01 12:00:00';
}
// or 
$arr = [
  'createdAt' => '2024-23-01 12:00:00'
];
```
now I want to normalize it with `$normalizer->normalize()`. How can I ensure that $createdAt is a valid datetime string of a given format (format returned from api can be different than this string format)? 
For object of class A I could write a custom normalizer and validate this string is a valid datetime - if valid then just return the string, if not then throw an exception. However this does not scale. I have to write custom normalizers for every class that has such a field.
Maybe there a way, like e.g. an attribute, to inform serializer to do that?
Or some other ideas?

In case of an array key there is nothing I can do.

**EDIT:**
> Maybe there a way, like e.g. an attribute, to inform serializer to do that?

Hmm, does normalizer even have access to the property's metadata? I can see that it just receives a raw value to normalize. If so then the only way for the normalizer is to work on the object and iterate over properties and look for that attribute.

**Comment by MatTheCat** [comment]
> How can I ensure that $createdAt is a valid datetime string of a given format (format returned from api can be different than this string format)?

Symfony provides [a `DateTime` constraint](https://symfony.com/doc/current/reference/constraints/DateTime.html). Also look at [the `Collection` constraint](https://symfony.com/doc/current/reference/constraints/Collection.html) if you need to validate arrays.

Normalization is another concept; you'd get a `NotNormalizableValueException` trying to normalize a datetime from an incorrect format.

**Reply by ghost** [reply]
> Normalization is another concept; you'd get a NotNormalizableValueException trying to normalize a datetime from an incorrect format.

You could have misunderstood me. I have a string property that holds a datetime value. Then, when an object that contains this property is going though the serialization process I would like to have this datetime string validated.
Maybe DateTimeNormalizer could support this validation but we would have to find a way other than the type of the value to determine whether normalizer supports this type (which is a string). Maybe via attribute but do we have access to this at this point?

```php
class A
{
  #[Serializer\Attribute\DateTimeString(format: ATOM)]
  public string $createdAt;
}
```
So ultimately, in this specific case the normalizer would not behave like a normalizer but a validator.

**Reply by MatTheCat** [reply]
Um yes I don't understand why you want a normalizer to act as a validator. Why not validate, *then* normalize?
